### PR TITLE
[ROCM] enabling gemm_rewriter subtests and several HLO passes for AMDGPU compiler

### PR DESCRIPTION
**Pull request by pemeliya**
We enable the majority of gemm_rewriter subtests on ROCM as well as activated DotDimensionMerger and PadForGemms 
HLO passes on ROCm.

The test xla/service/gpu/tests/gemm_rewrite_test.cc is very large. In order to easily enable/disable individual subtests for CUDA and ROCM paltforms, I have integrated the check function, which, in its most generic form, is implemented as follows:
```
  bool CudaOrRocmCheck(
          absl::AnyInvocable<bool(se::CudaComputeCapability)> cuda_fun,
          absl::AnyInvocable<bool(se::RocmComputeCapability)> rocm_fun) {
    const auto& gcomp = GpuComputeComp();            
    if(auto pcuda = std::get_if< se::CudaComputeCapability >(&gcomp); pcuda) {
      return (cuda_fun ? cuda_fun(*pcuda) : true); 
    } else {
      auto rocm = std::get< se::RocmComputeCapability >(gcomp);
      return (rocm_fun ? rocm_fun(rocm) : true); 
    }
```

Besides, I also override gtest ```void setUp()``` function to conditionally skip the whole test-suites based on the target architecture. The CUDA part is almost always not touched: only ```CudaCompileCapability``` is replaced with a generic variant ```GpuCompileCapability``` and respective checks.

@xla-rotation: could you please have a look ?

### CODE REVIEW

File: xla/service/gpu/amdgpu_compiler.cc [hunk 1]
```diff
@@ -116,10 +119,29 @@ Status AMDGPUCompiler::OptimizeHloPostLayoutAssignment(
...
+
+  HloPassPipeline pre_pipeline("AMDGPU post-layout_assignment part 1");
+
+  auto rocm_compute_capability = std::get<se::RocmComputeCapability>(
+      gpu_target_config.device_description.gpu_compute_capability());
+
+  pre_pipeline.AddPass<DotDimensionMerger>();
+
+  for (const auto& req : HipblasPaddingRequirements) {
+    //if (rocm_compute_capability.gfx_version() == '') {
```
Comment (ezhulenev) [inline xla/service/gpu/amdgpu_compiler.cc:131]: remove?

File: xla/service/gpu/cublas_padding_requirements.cc [hunk 2]
```diff
@@ -27,20 +27,27 @@ namespace gpu {
...
-                              const se::CudaComputeCapability cc) {
-  for (const CublasPaddingRequirement& requirement :
-       CublasPaddingRequirements) {
-    if (cc.IsAtLeast(requirement.min_compute_capability) &&
-        data_type == requirement.data_type &&
-        size % requirement.multiple_of != 0) {
-      return true;
+                              const se::GpuComputeCapability cc) {
+
+  if (auto pcc = std::get_if<se::CudaComputeCapability>(&cc); pcc) {
```
Comment (ezhulenev) [inline xla/service/gpu/cublas_padding_requirements.cc:32]: `auto* pcc` to make it clear that it's a pointer

even better use a std::visit here

```
struct RequiresPaddingVisitor {
  bool operator(se::CudaComputeCapability cc) {}
}

return std::visit(RequiresPaddingVisitor{}, cc);
```

File: xla/service/gpu/cublas_padding_requirements.h [hunk 3]
```diff
@@ -31,15 +31,25 @@ struct CublasPaddingRequirement {
   int multiple_of;
 };
 
+struct HipblasPaddingRequirement {
+  // std::string_view gpu_arch;
```
Comment (ezhulenev) [inline xla/service/gpu/cublas_padding_requirements.h:35]: remove?

File: xla/service/gpu/gemm_rewriter_triton.cc [hunk 4]
```diff
@@ -1279,9 +1279,14 @@ FusionContext::GetPropagatedDimOrdersAndRequirementsIfProfitablyFusible(
       std::move(std::get<DimOrdersAndReqs>(result_or_error));
   int fusion_level =
       hlo.GetModule()->config().debug_options().xla_gpu_triton_fusion_level();
-  if (!std::get<se::CudaComputeCapability>(gpu_version)
-           .IsAtLeast(se::CudaComputeCapability::AMPERE)) {
-    fusion_level = std::min(fusion_level, 1);
+  if (auto pcuda = std::get_if<se::CudaComputeCapability>(&gpu_version); pcuda) {
```
Comment (ezhulenev) [inline xla/service/gpu/gemm_rewriter_triton.cc:1282]: formatting is of, and again std visitor would be better here

**Comment by ezhulenev** [comment]
we dispatch to CUDA/ROCM CC many time here, I'd suggest to add overload pattern to gemm_rewriter and replace all if/else with an std::visit (https://www.modernescpp.com/index.php/visiting-a-std-variant-with-the-overload-pattern/)

File: xla/service/gpu/cublas_padding_requirements.cc (hunk 2 above)
Comment (pemeliya) [reply xla/service/gpu/cublas_padding_requirements.cc]: yes, std::visit is a good idea, indeed !

**Review by ezhulenev — APPROVED** [review]

**Comment by ddunl** [comment]
I think needs rebase before it can be imported (+ CI fix)

**Review by akuegel — APPROVED** [review]

**Comment by pemeliya** [comment]
> I think needs rebase before it can be imported (+ CI fix)

fixed! there was a missing const reference in triton function after rebase

**Review by akuegel — APPROVED** [review]

**Review by akuegel — APPROVED** [review]

File: xla/service/gpu/tests/gemm_rewrite_test.cc [hunk 5]
```diff
@@ -20,6 +20,7 @@ limitations under the License.
 
 #include <gtest/gtest.h>
 #include "absl/strings/str_replace.h"
+#include "absl/functional/any_invocable.h"
```
[the remark this answers is not in the document]
Comment (pemeliya) [reply xla/service/gpu/tests/gemm_rewrite_test.cc:23]: suprisingly, bazel did not complain about this missing dependency, but I have added it anyway

**Review by akuegel — APPROVED** [review]

**Comment by pemeliya** [comment]
I wonder, if there anything still needs to be done on this PR ? Thanks!

**Comment by akuegel** [comment]
> I wonder, if there anything still needs to be done on this PR ? Thanks!

Waiting for internal approval by someone else than me. In our process, the one who approves on GitHub cannot also approve internally.

@xla-rotation PTAL

**Pull request closed without merging**